File: 프로그래머스/2/42584. 주식가격/주식가격.py
def solution(prices):
    n = len(prices)
    answer = [0] * n 
    stack = []
    
    for i in range(n):
        # 가격이 떨어지는 시점 찾으면 그만큼 시간 계산
        while stack and prices[i] < prices[stack[-1]]:
            # 떨어지는 시점
            j = stack.pop()
            answer[j] = i - j 
        stack.append(i)
    
    while stack:
        j = stack.pop()
        answer[j] = n - 1 - j
        
    return answer

File: 프로그래머스/2/42584. 주식가격/주식가격.java
import java.util.ArrayDeque;

class Solution {
    public int[] solution(int[] prices) {
        ArrayDeque<Integer> stack = new ArrayDeque<>();
        int[] result = new int[prices.length];
        
        for(int i = 0; i < prices.length; i++){
              while(!stack.isEmpty() && prices[stack.peekLast()] > prices[i]){
                    // 떨어진 경우
                    int topIndex = stack.removeLast();
                    result[topIndex] = i - topIndex;
                }
            
            stack.add(i);
        }
        
        while(!stack.isEmpty()){
            int topIndex = stack.removeLast();
            result[topIndex] = prices.length - 1 - topIndex;
        }
        
        return result;
    }
}